### ISSUE TITLE
Swap docker.io databases for Quay ones for testing

### ISSUE THREAD

**Issue by Karm** [quarkusio/quarkus#40960]
### Description

I noticed GHA CI failing again with DockerHub issues, I guess quota or unavailability. 

```
[INFO] Quarkus - Integration Tests - Hibernate Reactive - MariaDB FAILURE [ 16.424 s]
[INFO] Quarkus - Integration Tests - Hibernate Reactive - MySQL FAILURE [ 16.143 s]
...
Failed to execute goal io.fabric8:docker-maven-plugin:0.44.0:build (docker-start) on project 
quarkus-integration-test-hibernate-reactive-mariadb: Unable to pull 'docker.io/mariadb:10.11' 
from registry 'docker.io' : {"message":"Get \"https://registry-1.docker.io/v2/\": net/http: 
request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"} 
(Internal Server Error: 500) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.fabric8:docker-maven-plugin:0.44.0:build
(docker-start) on project quarkus-integration-test-hibernate-reactive-mariadb: Unable to pull 'docker.io/mariadb:10.11' 
from registry 'docker.io' : {"message":"Get \"https://registry-1.docker.io/v2/\": net/http: request canceled while waiting
for connection (Client.Timeout exceeded while awaiting headers)"} (Internal Server Error: 500
```

### Implementation ideas

I had similar issues before in my CI and I fixed those by migrating tests to use e.g. 
```
quay.io/debezium/postgres
and
quay.io/mariadb-foundation/mariadb-devel
```

@yrodiere ?

**Comment by yrodiere** [comment]
I wouldn't recommend using `debezium/postgres` as I'm not sure what's in there, and I don't know where it's headed; IMO tests should run on a vanilla postgres container, or at the very least an official one.

`mariadb-foundation/mariadb-devel`, similarly, seems different than what people would use in production; I don't know to what extent.

That being said, I haven't seen so many Docker issues in other builds... Only Quarkus. Hibernate ORM builds (also using GitHub Actions) seem to work fine most of the time... Maybe the Quarkus build is hitting docker.io too hard?

If dockerhub is hard to reach, we should probably investigate caching proxies or mirrors instead. 

I know that for the part of Hibernate CI that runs on Jenkins, we use a private Nexus server for Maven dependencies; the default configuration handles Maven Central just fine, we just have to host a container somewhere.
I also know Nexus is able to proxy a docker registry: https://help.sonatype.com/en/proxy-repository-for-docker.html

Maybe we should investigate using a Nexus proxy in the Quarkus build as well? Of course it would have to be an option we set exclusively in Github actions, and since it runs on the cloud (aka "somebody else's computer"), we would have to rely on secrets... which probably means it wouldn't work for pull requests :/

**Comment by yrodiere** [comment]
> Could we cache the Docker images in the GitHub Actions cache? Problem is that we will probably hit the cache limit...

Yes we could, and yes the docker limit would be a problem. That being said, I've mentioned before we're currently using the cache for very dubious things and could get back a bit of space: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Persistent.20Develocity.20build.20cache.20in.20PRs/near/435358207

> oh, actually yes, there are suspicious entries about gradle setup on pull requests
>
> also, m2 repo cache on PRs... we don't want that, do we?
>
> I mean ideally we do, but it's 1GB, so not reasonable to cache for every PR. I'd expect the local build cache to be a much more efficient use of our build cache, since it could remove the need of downloading some of these dependencies entirely.
> 
> @gsmet WDYT of:
>
> * preventing cache writes for Maven dependencies / gradle setup on PRs. We'd only write to the cache from branch builds, restricting ourselves to read on PRs (so PRs benefit from the main branch's cache). This would save... close to half of our current cache usage, around 4GB.
> * and enabling cache read/write for the Develocity local build cache on PRs? This hopefully would take less that 4GB.
> FWIW cache writes for the Maven repo on PRs, if they are not just a fluke, may be a big problem right now, one that we didn't notice but that may be causing cache evictions for more important caches on the main branch, leading to cache misses for all builds (main or PR
>
> > Yoann Rodière [said](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Persistent.20Develocity.20build.20cache.20in.20PRs/near/435359156):
> >
> > preventing cache writes for Maven dependencies / gradle setup on PRs
> There's a lookup-only option on the cache action that does just that, from what I can see.
>
> Let me know what you think, I'll try sending a PR if you're interested. Might try that on Hibernate ORM as well.

**Comment by Karm** [comment]
Not a database, but another one:

```
Caused by: org.testcontainers.containers.ContainerFetchException: 
Can't get Docker image: RemoteDockerImage(imageName=axllent/mailpit:latest, imagePullPolicy=DefaultPullPolicy(),
...
Caused by: com.github.dockerjava.api.exception.DockerException: 
Status 429: {"message":"toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"}
```

**Comment by ppalaga** [comment]
We are also hitting the dockerhub limits on an internal Jenkins instance.

I see two options that would help us to switch to a proxying container registry:

1. Remove `docker.io/` from the default image names in https://github.com/quarkusio/quarkus/blob/main/build-parent/pom.xml#L78-L94 so that testcontainers', `hub.image.name.prefix=mirror.gcr.io/` or `TESTCONTAINERS_HUB_IMAGE_NAME_PREFIX=mirror.gcr.io/` starts working - see https://java.testcontainers.org/features/image_name_substitution/#:~:text=Testcontainers%20will%20not,com%20host%20part

2. Implement a `org.testcontainers.utility.ImageNameSubstitutor` that would allow substituting `docker.io` for something else via Quarkus configuration.

While 1. would be easy and fast, 2. could allow finer grained substitutions. 

I'd vote for going with 1.

**Comment by ppalaga** [comment]
A PR implementing the option 1: https://github.com/quarkusio/quarkus/pull/49969

**Comment by yrodiere** [comment]
Note if you're using Jenkins, there are infrastructure solutions: you don't need to change project code, you can just configure docker/podman to use mirrors.

@beikov and @marko-bekhta will know best, but here is what we do for Hibernate's CI: https://github.com/hibernate/hibernate-infra/blob/main/roles/jenkins-worker/files/etc/docker/daemon.json#L4

EDIT: I suppose this can be done on GitHub Actions as well, but I kind of expected GitHub to have its own mirrors in place? Might be wrong.

**Comment by ppalaga** [comment]
Thanks for the feedback @yrodiere! The podman asking for repo clearly makes the option 1 a no-go.
